File: breathing-animation/cli/python/breathing.py
import time
import sys
import math
import signal

# SWEBOK v4 "Serene Palette" (TrueColor RGB)
COLOR_RESET = "\033[0m"
# Inhale: Emerald #34d399 (52, 211, 153)
COLOR_EMERALD = "\033[38;2;52;211;153m"
# Hold: Blue #60a5fa (96, 165, 250)
COLOR_BLUE = "\033[38;2;96;165;250m"
# Exhale: Rose #fb7185 (251, 113, 133)
COLOR_ROSE = "\033[38;2;251;113;133m"

HIDE_CURSOR = "\033[?25l"
SHOW_CURSOR = "\033[?25h"
CLEAR_SCREEN = "\033[2J\033[H"
MOVE_HOME = "\033[H"

TECHNIQUES = {
    "1": {
        "name": "Box Breathing",
        "phases": [
            ("Inhale", 4.0, COLOR_EMERALD),
            ("Hold", 4.0, COLOR_BLUE),
            ("Exhale", 4.0, COLOR_ROSE),
            ("Hold", 4.0, COLOR_BLUE)
        ]
    },
    "2": {
        "name": "Diaphragmatic Breathing",
        "phases": [
            ("Inhale", 5.0, COLOR_EMERALD),
            ("Exhale", 5.0, COLOR_ROSE)
        ]
    },
    "3": {
        "name": "Alternate Nostril",
        "phases": [
            ("Inhale Left", 4.0, COLOR_EMERALD),
            ("Hold", 4.0, COLOR_BLUE),
            ("Exhale Right", 4.0, COLOR_ROSE),
            ("Hold", 4.0, COLOR_BLUE),
            ("Inhale Right", 4.0, COLOR_EMERALD),
            ("Hold", 4.0, COLOR_BLUE),
            ("Exhale Left", 4.0, COLOR_ROSE),
            ("Hold", 4.0, COLOR_BLUE)
        ]
        ]
    },
    "4": {
        "name": "4-7-8 Relaxing Breath",
        "phases": [
            ("Inhale", 4.0, COLOR_EMERALD),
            ("Hold", 7.0, COLOR_BLUE),
            ("Exhale", 8.0, COLOR_ROSE)
        ]
    }
}

def signal_handler(sig, frame):
    sys.stdout.write(SHOW_CURSOR + COLOR_RESET + "\nNamaste.\n")
    sys.exit(0)

signal.signal(signal.SIGINT, signal_handler)

def draw_frame(phase_name, progress, color, duration):
    # Move to top left
    sys.stdout.write(MOVE_HOME)
    
    # Header
    sys.stdout.write(f"\nMindful Breathing Visualizer (Python TUI v2.0)\n\n")

    # Phase Status
    remaining = duration * (1.0 - progress)
    sys.stdout.write(f"Phase: {phase_name:<15} ({remaining:5.1f}s)     \n\n")

    # Progress Bar
    bar_width = 40
    
    # Visual Progress Calculation (Expand on Inhale, Contract on Exhale)
    visual_progress = progress
    if "Inhale" in phase_name:
        visual_progress = progress
    elif "Exhale" in phase_name:
        visual_progress = 1.0 - progress
    else: 
         # Hold logic: if holding after inhale, stay full. 
         # Note: Simple stateless hold logic for now involves just checking the name
         # Ideally we'd pass previous state, but this heuristic works for standard techniques
         # If we just finished Inhale, we are Full. If Exhale, Empty.
         # This simple script assumes Hold is usually "Full" in Box Breathing contexts shown here for simplicity,
         # or we can just keep it static at 50% or full. 
         # Let's keep it simple: 
         visual_progress = 1.0 # Default to "Full" hold for visual stability

    current_width = int(visual_progress * bar_width)
    bar_str = "=" * current_width
    
    sys.stdout.write(f"      {color}[{bar_str:<{bar_width}}]{COLOR_RESET}\n\n")

    # Breathing Circle (Text Scale)
    scale = 1.0 + (visual_progress * 2.0)
    dots = int(scale * 5)
    lung_str = "( " + ("●" * dots) + " )"
    
    sys.stdout.write(f"       {color}{lung_str:<30}{COLOR_RESET}\n")
    
    sys.stdout.flush()

def run_phase(phase):
    name, duration, color = phase
    fps = 60
    frame_delay = 1.0 / fps
    start_time = time.time()

    while True:
        now = time.time()
        elapsed = now - start_time
        
        if elapsed >= duration:
            # SWEBOK KA 2 Audio Feedback (System Beep)
            sys.stdout.write("\a")
            sys.stdout.flush()
            break
            
        progress = elapsed / duration
        draw_frame(name, progress, color, duration)
        time.sleep(frame_delay)

def main():
    sys.stdout.write(CLEAR_SCREEN)
    print("Mindful Breathing Visualizer (Python TUI v2.0)")
    print("1. Box Breathing")
    print("2. Diaphragmatic Breathing")
    print("3. Alternate Nostril Breathing")
    print("4. 4-7-8 Relaxing Breath")
    choice = input("Select a technique (1-4): ").strip()
    
    technique = TECHNIQUES.get(choice, TECHNIQUES["1"])
    phases = technique["phases"]

    sys.stdout.write(CLEAR_SCREEN + HIDE_CURSOR)

    try:
        while True:
            for phase in phases:
                run_phase(phase)
    except KeyboardInterrupt:
        sys.stdout.write(SHOW_CURSOR + COLOR_RESET + "\nNamaste.\n")

if __name__ == "__main__":
    main()

File: breathing-animation/cli/go/breathing.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

// SWEBOK v4 "Serene Palette" (TrueColor RGB)
const (
	ColorReset   = "\033[0m"
	ColorEmerald = "\033[38;2;52;211;153m"  // Inhale
	ColorBlue    = "\033[38;2;96;165;250m"  // Hold
	ColorRose    = "\033[38;2;251;113;133m" // Exhale
)

const (
	HideCursor  = "\033[?25l"
	ShowCursor  = "\033[?25h"
	ClearScreen = "\033[2J\033[H"
	MoveHome    = "\033[H"
)

type Phase struct {
	Name     string
	Duration time.Duration // internal duration
	Color    string
}

func main() {
	fmt.Print(ClearScreen)
	fmt.Println("Mindful Breathing Visualizer (Go TUI v3.1.0)")
	fmt.Println("1. Box Breathing")
	fmt.Println("2. Diaphragmatic Breathing")
	fmt.Println("3. Alternate Nostril Breathing")
	fmt.Print("Select a technique (1-3): ")

	reader := bufio.NewReader(os.Stdin)
	choice, _ := reader.ReadString('\n')
	choice = strings.TrimSpace(choice)

	fmt.Print(ClearScreen)
	fmt.Println("Mindful Breathing Visualizer (Go TUI v3.1.0)")
	fmt.Println("1. Circle (●)")
	fmt.Println("2. Square (■)")
	fmt.Println("3. Lotus (🪷)")
	fmt.Println("4. Star (★)")
	fmt.Println("5. Flower (✿)")
	fmt.Println("6. Hexagon (⬢)")
	fmt.Println("7. Turtle (🐢)")
	fmt.Print("Select a shape (1-7): ")

	shapeChoice, _ := reader.ReadString('\n')
	shapeChoice = strings.TrimSpace(shapeChoice)

	shapeChar := "●"
	switch shapeChoice {
	case "2": shapeChar = "■"
	case "3": shapeChar = "🪷"
	case "4": shapeChar = "★"
	case "5": shapeChar = "✿"
	case "6": shapeChar = "⬢"
	case "7": shapeChar = "🐢"
	}

	var phases []Phase

	switch choice {
	case "2":
		phases = []Phase{
			{"Inhale", 5 * time.Second, ColorEmerald},
			{"Exhale", 5 * time.Second, ColorRose},
		}
	case "3":
		phases = []Phase{
			{"Inhale Left", 4 * time.Second, ColorEmerald},
			{"Hold", 4 * time.Second, ColorBlue},
			{"Exhale Right", 4 * time.Second, ColorRose},
			{"Hold", 4 * time.Second, ColorBlue},
			{"Inhale Right", 4 * time.Second, ColorEmerald},
			{"Hold", 4 * time.Second, ColorBlue},
			{"Exhale Left", 4 * time.Second, ColorRose},
			{"Hold", 4 * time.Second, ColorBlue},
		}
	default:
		phases = []Phase{
			{"Inhale", 4 * time.Second, ColorEmerald},
			{"Hold", 4 * time.Second, ColorBlue},
			{"Exhale", 4 * time.Second, ColorRose},
			{"Hold", 4 * time.Second, ColorBlue},
		}
	}

	fmt.Print(ClearScreen + HideCursor)

	// Main Loop
	for {
		for _, p := range phases {
			runPhase(p, shapeChar)
		}
	}
}

func runPhase(p Phase, shapeChar string) {
	fps := 60
	frameDelay := time.Second / time.Duration(fps)
	startTime := time.Now()

	// SWEBOK KA 2 Audio Feedback
	fmt.Print("\a")

	for {
		elapsed := time.Since(startTime)
		if elapsed >= p.Duration {
			break
		}

		progress := float64(elapsed) / float64(p.Duration)
		drawFrame(p, progress, elapsed, shapeChar)
		time.Sleep(frameDelay)
	}
}

func drawFrame(p Phase, progress float64, elapsed time.Duration, shapeChar string) {
	fmt.Print(MoveHome)
	fmt.Printf("\nMindful Breathing Visualizer (Go TUI v3.1.0)\n\n")

	remaining := p.Duration.Seconds() - elapsed.Seconds()
	fmt.Printf("Phase: %-15s (%0.1fs)     \n\n", p.Name, remaining)

	// Visual Logic
	visualProgress := progress
	if strings.Contains(p.Name, "Exhale") {
		visualProgress = 1.0 - progress
	} else if strings.Contains(p.Name, "Hold") {
		visualProgress = 1.0 // Simple heuristic
	}

	// Bar
	barWidth := 40
	fillWidth := int(visualProgress * float64(barWidth))
	bar := strings.Repeat("=", fillWidth)
	fmt.Printf("      %s[%-40s]%s\n\n", p.Color, bar, ColorReset)

	// Lung
	scale := 1.0 + (visualProgress * 2.0)
	dots := int(scale * 5.0)
	lungStr := "( " + strings.Repeat(shapeChar, dots) + " )"
	fmt.Printf("       %s%-30s%s\n", p.Color, lungStr, ColorReset)
}
